[3.1.0] Add Postgres Support

lupinitylabs opened the pull request:
This PR adds support for PostgreSQL databases, and additionally replaces the manual MySQL dump import logic with the default SchemaState `load()` method.

Also, it now prints the full dump path on the console after `protector:export`, and colorizes the output for improved legibility.

Testing status:

- [x] Export works with MySQL
- [x] Import works with MySQL
- [x] Remote import works with MySQL
- [x] Export works with Postgres
- [x] Import works with Postgres
- [x] Remote import works with Postgres.

In replacement of the manual drop/re-create database logic, we are now wiping the database prior to importing the dump. This can be prevented by using the `--no-wipe` option in `protector:import`.

> [!NOTE]
> It should be discussed whether we should just drop the tables, or really drop views and types, too.
> Since they are potentially both part of the dump, it seems to make sense to wipe it all.

lupinitylabs commented:
Please note that no cleanup has been done in detail.
I suppose a lot of the MySQL exec handling code can be removed after adopting this.

lupinitylabs commented:
We might also want to test a Postgres setup in the workflow from here on, but I guess your new workflow setup doesn't allow for different databases at this point?

mszulik commented:
> We might also want to test a Postgres setup in the workflow from here on, but I guess your new workflow setup doesn't allow for different databases at this point?

You can add support for it if you want, the repository is public: https://github.com/cybex-gmbh/github-workflows
Otherwise I will try to add it as soon as I have time.

I also created an issue for this: https://github.com/cybex-gmbh/github-workflows/issues/11

gael-connan-cybex commented on src/Exceptions/FailedWipeException.php:
```diff
@@ -0,0 +1,14 @@
+use Exception;
+
+/**
+ * Class FailedShellCommandException
```
pls update

gael-connan-cybex commented on src/Exceptions/FailedImportException.php:
```diff
@@ -0,0 +1,14 @@
+use Exception;
+
+/**
+ * Class FailedShellCommandException
```
pls update

lupinitylabs replied:
fixed

gael-connan-cybex commented on src/Traits/HasConfiguration.php:
```diff
@@ -21,6 +21,12 @@ trait HasConfiguration
     protected bool $createDb = true;
 
+    /**
+     * Defines whether dumps should drop existing databases before creating them.
```
the dump drops the db?

lupinitylabs replied:
updated.

lupinitylabs commented:
This pull request includes several updates to the `protector` package, notably adding PostgreSQL support, updating documentation, and enhancing the codebase with new features and improvements. The most important changes include updates to the `CHANGELOG.md`, improvements to the `README.md`, additions to the `composer.json` file, new schema state proxy classes, and modifications to the `Protector` class to handle PostgreSQL.

### Documentation Updates:
* [`CHANGELOG.md`](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edL3-R45): Updated to include PostgreSQL support and other version changes.
* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L23-R27): Updated supported databases section to include PostgreSQL and added a migration guide from v2.x to v3.x. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L23-R27) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R276-R279)

### Dependency and Configuration Updates:
* [`composer.json`](diffhunk://#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34R41): Added `laravel/pint` as a development dependency and included a new developer in the `authors` section. [[1]](diffhunk://#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34R41) [[2]](diffhunk://#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34R21-R25)
* [`pint.json`](diffhunk://#diff-2c2d15bc7fdad73f20daa8a343b8d93018ae6a8ae050cd4ede67719f1a5a5353R1-R38): Added configuration for `laravel/pint` code style fixer.

### Schema State Proxy Classes:
* [`src/Classes/AbstractPostgresSchemaStateProxy.php`](diffhunk://#diff-7198fc3e305cd27421cd152f8e822f592dc3e04e89b2e0b9a2ba29223944d7adR1-R62): Introduced an abstract class for PostgreSQL schema state proxy.
* [`src/Classes/PostgresSchemaStateProxy.php`](diffhunk://#diff-465ea6d569630fa3ad87026645769f6e2f71f7384ff5438ba2d283fa6939c869R1-R50): Added a concrete implementation for PostgreSQL schema state proxy.

### Core Functionality Enhancements:
* [`src/Protector.php`](diffhunk://#diff-c2d6c0150977bc65bc04cda9276be7517946ab40327d95bd102e586cb6d86238L65-R72): Updated to handle PostgreSQL schema state and added error handling for import and wipe operations. [[1]](diffhunk://#diff-c2d6c0150977bc65bc04cda9276be7517946ab40327d95bd102e586cb6d86238L65-R72) [[2]](diffhunk://#diff-c2d6c0150977bc65bc04cda9276be7517946ab40327d95bd102e586cb6d86238L88-R109)

mszulik commented:
> > We might also want to test a Postgres setup in the workflow from here on, but I guess your new workflow setup doesn't allow for different databases at this point?
> 
> You can add support for it if you want, the repository is public: https://github.com/cybex-gmbh/github-workflows Otherwise I will try to add it as soon as I have time.
> 
> I also created an issue for this: [cybex-gmbh/github-workflows#11](https://github.com/cybex-gmbh/github-workflows/issues/11)

Workflow update PR: https://github.com/cybex-gmbh/github-workflows/pull/15
Protector PR: https://github.com/cybex-gmbh/laravel-protector/pull/93

gael-connan-cybex commented on src/Classes/AbstractPostgresSchemaStateProxy.php:
```diff
@@ -0,0 +1,62 @@
+        $this->schemaState->load(...func_get_args());
+    }
+
+    protected function baseVariables(mixed $config): array
```
docblock

gael-connan-cybex replied:
will be cleaned up later

gael-connan-cybex reviewed (approved):
looks ok, has been in production already
